File: conversions/mime-types/elide-mime-types.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-mime-types

This demonstrates calling the TypeScript MIME types implementation from Python
using Elide's polyglot capabilities.

Benefits:
- One MIME database shared across TypeScript and Python
- Consistent file type detection across services
- No Python mime library needed
- Guaranteed type consistency
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Python support is ready

# Assuming Elide provides something like:
# from elide import require
# mime = require('./elide-mime-types.ts')

print("=== Python Calling TypeScript MIME Types ===\n")

# Example 1: Lookup MIME type by extension
# mime_type = mime.lookup('document.pdf')
# print(f"PDF MIME type: {mime_type}")

# Example 2: File Upload Handler (Flask)
# from flask import Flask, request
#
# @app.route('/upload', methods=['POST'])
# def handle_upload():
#     file = request.files['file']
#     mime_type = mime.lookup(file.filename)
#
#     # Validate file type
#     allowed_types = ['image/jpeg', 'image/png', 'application/pdf']
#     if mime_type not in allowed_types:
#         return {'error': 'Invalid file type'}, 400
#
#     # Set correct Content-Type
#     ct = mime.contentType(file.filename)
#     return Response(file.read(), headers={'Content-Type': ct})

# Example 3: Static File Server
# def serve_static(file_path: str):
#     """Serve static file with correct Content-Type"""
#     content_type = mime.contentType(file_path)
#     with open(file_path, 'rb') as f:
#         return {'body': f.read(), 'headers': {'Content-Type': content_type}}

# Example 4: File Type Validation
# def validate_upload_type(filename: str) -> bool:
#     """Validate file is an allowed type"""
#     mime_type = mime.lookup(filename)
#     allowed = ['image/jpeg', 'image/png', 'image/gif', 'application/pdf']
#     return mime_type in allowed

print("Real-world use case:")
print("- Python API determines Content-Type for file uploads")
print("- Uses same TypeScript implementation as Node.js service")
print("- Guarantees consistent MIME types across storage services")
print("- No need to install Python mimetype library")
print()

print("Example: File Storage Service")
print("┌─────────────────────────────────────────────┐")
print("│   Elide MIME Types (TypeScript)            │")
print("│   elide-mime-types.ts                      │")
print("└─────────────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │Upload  │          │Storage │")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same MIME detection everywhere!")
print()

print("Problem Solved:")
print("Before: Python mimetypes + JavaScript mime = different types")
print("After: One Elide implementation = 100% consistent types")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  mime = require('./elide-mime-types.ts')")
print("  mime_type = mime.lookup('file.pdf')  # That's it!")

File: converted/utilities/mime-types/elide-mime-types.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-mime-types
#
# This demonstrates calling the TypeScript MIME types implementation from Ruby
# using Elide's polyglot capabilities.
#
# Benefits:
# - One MIME database shared across TypeScript and Ruby
# - Consistent file type detection across microservices
# - No Ruby mime-types gem needed
# - Perfect for Rails, Sinatra file handling

puts "=== Ruby Calling TypeScript MIME Types ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# mime = Elide.require('./elide-mime-types.ts')

# Example 1: Rails File Upload
# class UploadsController < ApplicationController
#   def create
#     file = params[:file]
#     mime_type = mime.lookup(file.original_filename)
#
#     # Validate file type
#     unless ['image/jpeg', 'image/png', 'application/pdf'].include?(mime_type)
#       return render json: { error: 'Invalid file type' }, status: 400
#     end
#
#     # Store with correct MIME type
#     storage.save(file, mime_type: mime_type)
#   end
# end

# Example 2: Static File Server (Sinatra)
# require 'sinatra'
#
# get '/files/:filename' do
#   file_path = File.join(PUBLIC_DIR, params[:filename])
#   content_type mime.contentType(file_path)
#   send_file file_path
# end

# Example 3: File Type Detection
# def detect_file_type(filename)
#   mime_type = mime.lookup(filename)
#   is_image = mime_type&.start_with?('image/')
#   is_document = ['application/pdf', 'application/msword'].include?(mime_type)
#   { type: mime_type, image: is_image, document: is_document }
# end

puts "Real-world use case:"
puts "- Ruby Rails app determines MIME types for uploads"
puts "- Uses same TypeScript implementation as Node.js CDN"
puts "- Guarantees consistent Content-Type headers"
puts "- No mime-types gem needed"
puts ""

puts "Example: CDN Service"
puts "┌─────────────────────────────────────────────┐"
puts "│   Elide MIME Types (TypeScript)            │"
puts "│   elide-mime-types.ts                      │"
puts "└─────────────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Ruby  │  │ Rails  │"
puts "    │  CDN   │  │ Worker │  │  App   │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same MIME types everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Different MIME databases = inconsistent types"
puts "After: One Elide implementation = perfect consistency"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/mime.rb"
puts "  MIME = Elide.require('./elide-mime-types.ts')"
puts "  "
puts "  # app/controllers/uploads_controller.rb"
puts "  def create"
puts "    mime_type = MIME.lookup(params[:file].original_filename)"
puts "  end"
